src/main/java/ui/BoardNameDialog.java

```diff
@@ -0,0 +1,150 @@
+package ui;
+
+import java.util.ArrayList;
+
+import prefs.Preferences;
+import javafx.application.Platform;
+import javafx.geometry.Insets;
+import javafx.geometry.Pos;
+import javafx.scene.control.ButtonBar;
+import javafx.scene.control.ButtonType;
+import javafx.scene.control.Dialog;
+import javafx.scene.control.Button;
+import javafx.scene.control.TextField;
+import javafx.scene.layout.GridPane;
+import javafx.scene.layout.HBox;
+import javafx.scene.layout.Region;
+import javafx.scene.text.Text;
+import javafx.stage.Modality;
+import javafx.stage.Stage;
+
+public class BoardNameDialog extends Dialog<String> {
+    
+    Preferences prefs;
+    TextField nameField;
+    Text prompt;
+    Text errorText;
+    Button submitButton;
+
+    private ArrayList<String> invalidNames = new ArrayList<>(); 
+    
+    private String previousText = "";
+    private static final String DEFAULT_NAME = "New Board";
+    private static final int BOARD_MAX_NAME_LENGTH = 100;
+    private static final String ERROR_DUPLICATE_NAME = "Warning: duplicate name. Overwrite?";
+    private static final String ERROR_EMPTY_NAME = "Error: empty name.";
+    private static final String ERROR_LONG_NAME = "Error: board name cannot exceed %d letters.";
+    private static final String ERROR_INVALID_NAME = "Error: invalid name.";
+
+    public BoardNameDialog(Preferences prefs, Stage mainStage) {
+        this.prefs = prefs;
+        initializeDialog(mainStage);
+        setupGrid();
+        createButtons();
+        addListener();
+        
+        invalidNames.add("none"); // TODO possibly extend the list in the future
+    }
+    
+    private void initializeDialog(Stage mainStage) {
+        initOwner(mainStage);
+        initModality(Modality.APPLICATION_MODAL);
+        setTitle("Save board as");
+    }
+    
+    private void createButtons() {
+        ButtonType submitButtonType = new ButtonType("Save", ButtonBar.ButtonData.OK_DONE);
+        getDialogPane().getButtonTypes().addAll(submitButtonType, ButtonType.CANCEL);
+        submitButton = (Button) getDialogPane().lookupButton(submitButtonType);
+        
+        setResultConverter(submit -> {
+            if (submit == submitButtonType) {
+                return nameField.getText();
+            }
+            return null;
```

Comment:
It does as far as I know. I remember returning null because I can't return an empty string or "" and I had to search online.